### PR TITLE
[doc] Homogenize the format of the c++ docstrings and add more doxygen tags in some places

### PR DESCRIPTION
**Pull request by tmadlener**
BEGINRELEASENOTES
- Make all docstrings of c++ classes use the `///` style (instead of the previous mix of `///` and /**/`.
- Extend and improve existing docstrings (e.g. add `@params` and `@returns` tags, add information or reword).

ENDRELEASENOTES

### CODE REVIEW

File: include/podio/Frame.h [hunk 1]
```diff
@@ -66,14 +64,14 @@ std::optional<podio::CollectionReadBuffers> unpack(FrameDataT* data, const std::
   return data->getCollectionBuffers(name);
 }
 
-/**
- * Frame class that serves as a container of collection and meta data.
- */
+/// The Frame is a generalized (event) data container that aggregates all
+/// relevent data.
```
Comment (m-fila) [inline include/podio/Frame.h:68]: typo
```suggestion
/// relevant data.
```

File: include/podio/RNTupleWriter.h [hunk 2]
```diff
@@ -17,38 +17,88 @@
 
 namespace podio {
 
+/// The RNTupleWriter writes podio files into ROOT files using the new RNTuple
+/// format.
+///
+/// Each category get's its own RNTuple. Additionally, there is a podio_metadata
```
Comment (m-fila) [inline include/podio/RNTupleWriter.h:23]: typo
```suggestion
/// Each category gets its own RNTuple. Additionally, there is a podio_metadata
```

File: include/podio/ROOTWriter.h [hunk 3]
```diff
@@ -21,47 +21,80 @@ class Frame;
 class CollectionBase;
 class GenericParameters;
 
+/// The ROOTWriter writes podio files into ROOT files using TTrees.
+///
+/// Each category get's its own TTree. Additionally, there is a podio_metadata
```
Comment (m-fila) [inline include/podio/ROOTWriter.h:26]: typo
```suggestion
/// Each category gets its own TTree. Additionally, there is a podio_metadata
```

File: include/podio/CollectionBufferFactory.h [hunk 4]
```diff
@@ -11,20 +11,19 @@
...
- *
- * It is implemented as a singleton, which is populated at the time a shared
- * datamodel library is loaded. It is assumed that that happens early on in the
- * startup of an application, such that only a single thread will access the
- * factory instance for registering datatypes. Since the necessary creation
- * functions are part of the core datamodel library, this should be very easy to
- * achieve by simply linking to that library. Once the factory is populated it
- * can be safely accessed from multiple threads concurrently to obtain buffers.
- */
+/// The CollectionBufferFactory allows to create buffers of known datatypes,
```
Comment (andresailer) [inline include/podio/CollectionBufferFactory.h:14]:
```suggestion
/// The CollectionBufferFactory allows one to create buffers of known datatypes,
```

File: include/podio/DatamodelRegistry.h [hunk 5]
```diff
@@ -70,59 +64,57 @@ class DatamodelRegistry {
...
-   * returned.
-   *
-   * @param name The name of the EDM that should be registered
-   * @param definition The datamodel definition from which this EDM has been
-   * generated in JSON format
-   * @param relationNames the names of the relations and vector members for all
-   * datatypes that are defined for this EDM
-   *
-   */
+  /// Register a datamodel return the index in the registry.
```
Comment (andresailer) [inline include/podio/DatamodelRegistry.h:100]:
```suggestion
  /// Register a datamodel and return its index in the registry.
```

File: include/podio/Frame.h [hunk 6]
```diff
@@ -151,117 +143,182 @@ class Frame {
...
   inline void putParameter(const std::string& key, std::initializer_list<T>&& values) {
     putParameter<std::vector<T>>(key, std::move(values));
   }
 
-  /** Retrieve parameters via key from the internal store. Return type will
-   * either by a const reference or a value depending on the desired type.
-   */
+  /// Retrieve parameters via key from the internal store.
+  ///
+  /// The return type will either by a const reference or a value depending on
```
Comment (andresailer) [inline include/podio/Frame.h:274]:
```suggestion
  /// The return type will either be a const reference or a value depending on
```

File: include/podio/Frame.h [hunk 7]
```diff
@@ -151,117 +143,182 @@ class Frame {
...
   inline std::vector<std::string> getParameterKeys() const {
     return m_self->parameters().getKeys<T>();
   }
 
-  /** Get all **currently** available collections (including potentially
-   * unpacked ones from raw data)
-   */
+  /// Get all **currently** available collection names.
+  ///
+  /// @returns The names of all collections, including the ones that might still
```
Comment (andresailer) [inline include/podio/Frame.h:308]:
```suggestion
  /// @returns The names of all collections, including those that might still
```

File: include/podio/GenericParameters.h [hunk 8]
```diff
@@ -66,16 +66,14 @@ namespace detail {
...
- *  int, float and string or vectors of these types.
- *  They can be used  to store (user) meta data that is
- *  run, event or collection dependent.
- *  (based on lcio::LCParameters)
- *
- * @author F. Gaede, DESY
- * @date Apr 2020
- */
-
+/// GenericParameters objects allow to store generic named parameters of type
```
Comment (andresailer) [inline include/podio/GenericParameters.h:69]:
```suggestion
/// GenericParameters objects allow one to store generic named parameters of type
```

**Review by andresailer — COMMENTED** [review]
some typos, some grammar. Haven't read everything yet.

File: include/podio/ROOTReader.h [hunk 9]
```diff
@@ -36,86 +36,103 @@ class CollectionIDTable;
...
+  ///
+  /// @returns The podio build version
   podio::version::Version currentFileVersion() const {
     return m_fileVersion;
   }
 
-  /// Get the names of all the available Frame categories in the current file(s)
+  /// Get the names of all the available Frame categories in the current file(s).
+  ///
+  /// @returns The names of the available categores from the file
```
Comment (andresailer) [inline include/podio/ROOTReader.h:111]:
```suggestion
  /// @returns The names of the available categories from the file
```

File: include/podio/ROOTWriter.h [hunk 10]
```diff
@@ -21,47 +21,80 @@ class Frame;
...
+  ///
+  /// NOTE: Existing files will be overwritten without warning.
+  ///
+  /// @param filename The path to the file that will be created.
   ROOTWriter(const std::string& filename);
+
+  /// ROOTWriter destructor
+  ///
+  /// This also takes care of writing all the necessary metadata in order to be
+  /// able to read files back again.
```
Comment (andresailer) [inline include/podio/ROOTWriter.h:43]:
```suggestion
  /// This also takes care of writing all the necessary metadata to read files back again.
```

File: include/podio/RelationRange.h [hunk 11]
```diff
@@ -5,10 +5,8 @@
 #include <vector>
 
 namespace podio {
-/**
- * A simple helper class that allows to return related objects in a way that
- * makes it possible to use the return type in a range-based for loop.
- */
+/// A simple helper class that allows to return related objects in a way that
```
Comment (andresailer) [inline include/podio/RelationRange.h:8]:
```suggestion
/// A simple helper class that allows one to return related objects in a way that
```

File: include/podio/SIOWriter.h [hunk 12]
```diff
@@ -14,23 +14,61 @@ namespace podio {
...
+  ///
+  /// NOTE: Existing files will be overwritten without warning.
+  ///
+  /// @param filename The path to the file that will be created.
   SIOWriter(const std::string& filename);
+
+  /// SIOWriter destructor
+  ///
+  /// This also takes care of writing all the necessary metadata in order to be
+  /// able to read files back again.
```
Comment (andresailer) [inline include/podio/SIOWriter.h:37]:
```suggestion
  /// This also takes care of writing all the necessary metadata to read files back again.
```

File: include/podio/utilities/TypeHelpers.h [hunk 13]
```diff
@@ -50,35 +50,29 @@ namespace det {
...
-   * This will also make root dictionary generation fail. GCC works without this
-   * small workaround and is standard compliant in this case, whereas clang is
-   * not.
-   */
+  /// Helper struct to turn a tuple of types into a tuple of a template of types, e.g.
+  ///
+  /// std::tuple<int, float> -> std::tuple<std::vector<int>, std::vector<float>>
+  /// if the passed template is std::vector
+  ///
+  /// NOTE: making the template template parameter to Template variadic because
```
Comment (andresailer) [inline include/podio/utilities/TypeHelpers.h:71]: Should we use [`\note`](https://www.doxygen.nl/manual/commands.html#cmdnote) in these cases?

File: include/podio/utilities/TypeHelpers.h [hunk 14]
```diff
@@ -153,101 +133,79 @@ namespace detail {
...
+  /// determine whether T is (or could be) a podio generated default (immutable)
+  /// handle.
   template <typename T>
   using hasMutable_t = typename T::mutable_type;
 
-  /**
-   * Detector for checking the existance of an object_type type member. Used to
-   * determine whether T is (or could be) a podio generated mutable handle.
-   */
+  /// Detector for checking the existance of an object_type type member. Used to
```
Comment (andresailer) [inline include/podio/utilities/TypeHelpers.h:176]:
```suggestion
  /// Detector for checking the existence of an object_type type member. Used to
```

File: include/podio/utilities/TypeHelpers.h [hunk 15]
```diff
@@ -153,101 +133,79 @@ namespace detail {
...
-   *
-   * NOTE: this basically just checks the existance of the mutable_type and
-   * object_type type members, so it is rather easy to fool this check if one
-   * wanted to. However, for our purposes we mainly need it for a static_assert
-   * to have more understandable compilation error message.
-   */
+  /// Variable template for determining whether type T is a podio generated
+  /// handle class.
+  ///
+  /// NOTE: this basically just checks the existance of the mutable_type and
```
Comment (andresailer) [inline include/podio/utilities/TypeHelpers.h:184]:
```suggestion
  /// NOTE: this basically just checks the existence of the mutable_type and
```

File: podioVersion.in.h [hunk 16]
```diff
@@ -33,11 +33,9 @@
 
 namespace podio::version {
 
-/**
- * Version class consisting of 3 16 bit unsigned integers to hold the major,
- * minor and patch version. Provides constexpr comparison operators that allow
- * to use this class in constexpr-if clauses.
- */
+/// Version class consisting of 3 16 bit unsigned integers to hold the major,
```
Comment (andresailer) [inline podioVersion.in.h:36]:
```suggestion
/// Version class consisting of three 16 bit unsigned integers to hold the major,
```

File: podioVersion.in.h [hunk 17]
```diff
@@ -33,11 +33,9 @@
...
+/// minor and patch version. Provides constexpr comparison operators that allow
+/// to use this class in constexpr-if clauses.
```
Comment (andresailer) [inline podioVersion.in.h:38]:
```suggestion
/// one to use this class in constexpr-if clauses.
```

File: include/podio/utilities/TypeHelpers.h (hunk 13 above)
Comment (tmadlener) [reply include/podio/utilities/TypeHelpers.h]: Yes, and done.

**Pull request merged**